fix: use pre-install helm hook to prepare rancher for turtles

salasberryfin opened the pull request:
**What this PR does / why we need it**:

Currently, installing Rancher Turtles in an existing Rancher Manager after disabling embedded CAPI, results in validating/mutating webhooks being left behind after the feature is disabled. If you try and apply a CAPI manifest you will get an error:

```
time="2023-10-03T14:18:36Z" level=error msg="error syncing 'cluster-fleet-local-local-1a3d67d0a899/clusters-clusters': handler bundle-deploy: failed to create  │
│ resource: Internal error occurred: failed calling webhook \"default.cluster.cluster.x-k8s.io\": failed to call webhook: Post \"https://capi-webhook-service.cat/ │
│ tle-provisioning-capi-system.svc:443/mutate-cluster-x-k8s-io-v1beta1-cluster?timeout=10s\": service \"capi-webhook-service\" not found, requeuing"
```

To avoid having to manually delete these webhooks and disable the feature flag, this PR adds a `pre-install` Helm hook that:
1. Disables embedded CAPI.
2. Deletes mutating/validating webhooks (creating a Service Account, Cluster Role and Cluster Role Binding).

This is not the fanciest of solutions but it allows to install Rancher Turtles just using `helm install` in an existing Rancher Manager. We can later change Rancher and remove this hook if we find a more robust option.

**Which issue(s) this PR fixes**:
Fixes #185 

**Special notes for your reviewer**:

When this gets merged, we'll need to update the docs accordingly.

**Checklist**:

- [x] squashed commits into logical changes
- [ ] includes documentation
- [ ] adds unit tests
- [ ] adds or updates e2e tests

salasberryfin commented:
We may need to install Rancher before Rancher Turtles in the E2E environment for the `Feature` kind to be available. Looks like this is triggering CI failures.

Danil-Grigorev commented:
[E2e](https://github.com/rancher-sandbox/rancher-turtles/pull/190) tests are passing with this change, so failure here could be ignored.

salasberryfin commented:
Thanks @Danil-Grigorev. Should we consider this PR superseded by yours #190 and close it?

Danil-Grigorev commented:
I don’t know. Can go both ways, you can cherry pick the commit #190 and the other one could be closed. Or this PR could be merged without e2e tests passing, and followed by #190 separately.

salasberryfin commented:
Let's merge this first and then #190 and keep them as independent PRs.

richardcase commented on charts/rancher-turtles/templates/pre-install-job.yaml:
```diff
@@ -0,0 +1,73 @@
+---
```
Could we have a value that can be used to enable or disable this? Perhaps in the values.yaml we can have something like this:

```yaml
features:
  cleanup-rancher-webhooks: true
```

I can then add to the values.yaml in this section to enable/delete the labels update (#178)

richardcase commented on charts/rancher-turtles/templates/pre-install-job.yaml:
```diff
@@ -0,0 +1,73 @@
+      serviceAccountName: pre-install-job
+      containers:
+        - name: rancher-webhook-cleanup
+          image: bitnami/kubectl
```
Rancher does something similar. Do you know if it uses the same image?

salasberryfin replied:
I was investigating a bit and Rancher uses this https://github.com/rancher/kubectl utility that directly runs the command you pass in the container and terminates it rather than having a bash session that you can run `kubectl` commands in. The only issue with this is that you must specify the `kubectl` version as tag so having custom images won't be really possible. I guess this is fine considering it is a Rancher-owned image but it may be worth a look.

richardcase commented on charts/rancher-turtles/templates/pre-install-job.yaml:
```diff
@@ -0,0 +1,73 @@
+      serviceAccountName: pre-install-job
+      containers:
+        - name: rancher-webhook-cleanup
+          image: bitnami/kubectl
```
Also, i think we'd want to have the image configurable in the values.yaml for air-gapped scenarios.

richardcase reviewed (commented):
In my head i was thinking of something actually in the Rancher Turtles binary but this is an interesting approach.

salasberryfin commented:
I added flags for disabling embedded capi and deleting rancher webhooks:
```bash
rancherTurtles:
  image: controller
  tag: v0.0.0
  imagePullPolicy: Never
  namespace: rancher-turtles-system
  managerArguments: {}
  imagePullSecrets: []
  features:
    disable-embedded-capi:
      enabled: true
    cleanup-rancher-webhook:
      enabled: true
      image: rancher/kubectl
```
Also, changed the deleting webhook `Job` to use `rancher/kubectl` image. This has to now be separated into two different jobs: one for each webhook deletion (mutating/validating).

richardcase reviewed (approved):
Looks good, thanks for adding the flags in the values.

Although there is a chart linrt failure.

Danil-Grigorev commented on charts/rancher-turtles/values.yaml:
```diff
@@ -5,6 +5,12 @@ rancherTurtles:
   managerArguments: {}
   imagePullSecrets: []
+  features:
+    disable-embedded-capi:
+      enabled: true
```
```suggestion
    embedded-capi:
      disabled: true
```

Danil-Grigorev commented on charts/rancher-turtles/values.yaml:
```diff
@@ -5,6 +5,12 @@ rancherTurtles:
+  features:
+    disable-embedded-capi:
+      enabled: true
+    cleanup-rancher-webhook:
+      enabled: true
```
```suggestion
   rancher-webhook:
      cleanup: true
```

Danil-Grigorev commented on .github/workflows/test_chart.yaml:
```diff
@@ -54,10 +56,23 @@ jobs:
+        run: helm install rancher rancher-stable/rancher --namespace cattle-system --create-namespace --set bootstrapPassword=rancheradmin --set replicas=1 --set hostname="e2e.dev.rancher" --set 'extraEnv[0].name=CATTLE_FEATURES' --set global.cattle.psp.enabled=false --version ${{ env.RANCHER_VERSION }} --wait
+
       - name: Run chart-testing (install)
         run: helm install rancher-turtles out/charts/rancher-turtles/ -n rancher-turtles-system --create-namespace --wait --set cluster-api-operator.cluster-api.enabled=false --set cluster-api-operator.enabled=false
```
```suggestion
        run: helm install rancher-turtles out/charts/rancher-turtles/ -n rancher-turtles-system --create-namespace --wait --set cluster-api-operator.cluster-api.enabled=false --set cluster-api-operator.enabled=false --set features. disable-embedded-capi.enabled=false --set features.cleanup-rancher-webhook.enabled=false
```
As this workflow represents the default flow of installing rancher-turtles in a fresh cluster, we might not need to bring rancher with it. Just making sure the chart is installable.

salasberryfin replied:
I understand your point but I would say, considering these flags are enable by default, we should test them as well to validate the chart.

Danil-Grigorev reviewed (dismissed):
Some minor suggestions, but LGTM as is.

Danil-Grigorev commented on .github/workflows/test_chart.yaml:
```diff
@@ -54,10 +56,23 @@ jobs:
       - name: Add local docker image
         run: kind load docker-image ${{ env.MANIFEST_IMG }}:${{ env.TAG }}
 
+      - name: Add cert-manager chart repo
+        run: helm repo add jetstack https://charts.jetstack.io
+
+      - name: Add rancher chart repo
+        run: helm repo add rancher-stable https://releases.rancher.com/server-charts/stable
+
+      - name: Install cert-manager
+        run: helm install cert-manager jetstack/cert-manager --namespace cert-manager --create-namespace --version ${{ env.CERT_MANAGER_VERSION }} --set installCRDs=true --wait
+
+      - name: Install Rancher
+        run: helm install rancher rancher-stable/rancher --namespace cattle-system --create-namespace --set bootstrapPassword=rancheradmin --set replicas=1 --set hostname="e2e.dev.rancher" --set 'extraEnv[0].name=CATTLE_FEATURES' --set global.cattle.psp.enabled=false --version ${{ env.RANCHER_VERSION }} --wait
+
```
```suggestion
```

In addition to https://github.com/rancher-sandbox/rancher-turtles/pull/192/files#r1348559835
